File: tests/python/test_mwdx.py
"""
Python binding tests for MWDX indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


try:
    from my_project import (
        mwdx,
        mwdx_batch,
        MwdxStream
    )
except ImportError:
    pytest.skip("MWDX module not available - run 'maturin develop' first", allow_module_level=True)


class TestMwdx:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_mwdx_partial_params(self, test_data):
        """Test MWDX with default parameters - mirrors check_mwdx_partial_params"""
        close = np.array(test_data['close'], dtype=np.float64)


        result = mwdx(close, 0.2)
        assert isinstance(result, np.ndarray)
        assert len(result) == len(close)

    def test_mwdx_accuracy(self, test_data):
        """Test MWDX matches expected values from Rust tests - mirrors check_mwdx_accuracy"""
        close = np.array(test_data['close'], dtype=np.float64)
        expected = EXPECTED_OUTPUTS['mwdx']


        result = mwdx(close, expected['default_params']['factor'])

        assert len(result) == len(close)


        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=1e-7,
            msg="MWDX last 5 values mismatch"
        )


        compare_with_rust('mwdx', result, 'close', expected['default_params'])


    def test_mwdx_zero_factor(self):
        """Test MWDX fails with zero factor - mirrors check_mwdx_zero_factor"""
        input_data = np.array([10.0, 20.0, 30.0], dtype=np.float64)

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            mwdx(input_data, 0.0)

    def test_mwdx_negative_factor(self):
        """Test MWDX fails with negative factor - mirrors check_mwdx_negative_factor"""
        input_data = np.array([10.0, 20.0, 30.0], dtype=np.float64)

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            mwdx(input_data, -0.5)

    def test_mwdx_nan_factor(self):
        """Test MWDX fails with NaN factor"""
        input_data = np.array([10.0, 20.0, 30.0], dtype=np.float64)

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            mwdx(input_data, np.nan)


    def test_mwdx_very_small_dataset(self):
        """Test MWDX with single data point - mirrors check_mwdx_very_small_dataset"""
        data = np.array([42.0], dtype=np.float64)

        result = mwdx(data, 0.2)
        assert len(result) == 1
        assert result[0] == 42.0

    def test_mwdx_empty_input(self):
        """Test MWDX with empty input"""
        data_empty = np.array([], dtype=np.float64)

        with pytest.raises(ValueError, match="No input data was provided"):
            mwdx(data_empty, 0.2)


    def test_mwdx_reinput(self, test_data):
        """Test MWDX with re-input of MWDX result - mirrors check_mwdx_reinput"""
        close = np.array(test_data['close'], dtype=np.float64)


        first_result = mwdx(close, 0.2)


        second_result = mwdx(first_result, 0.3)

        assert len(second_result) == len(first_result)


        assert np.all(np.isfinite(second_result))


    def test_mwdx_nan_handling(self, test_data):
        """Test MWDX handling of NaN values - mirrors check_mwdx_nan_handling"""
        close = np.array(test_data['close'], dtype=np.float64)

        result = mwdx(close, 0.2)

        assert len(result) == len(close)


        assert np.all(np.isfinite(result))


    def test_mwdx_batch(self, test_data):
        """Test MWDX batch computation"""
        close = np.array(test_data['close'], dtype=np.float64)


        batch_result = mwdx_batch(
            close,
            (0.1, 0.5, 0.1)
        )


        assert isinstance(batch_result, dict)
        assert 'values' in batch_result
        assert 'factors' in batch_result


        assert batch_result['values'].shape == (5, len(close))
        assert len(batch_result['factors']) == 5


        individual_result = mwdx(close, 0.1)
        batch_first = batch_result['values'][0]
        assert_close(batch_first, individual_result, atol=1e-9, msg="MWDX first combination mismatch")


    def test_mwdx_streaming(self, test_data):
        """Test MWDX streaming interface - mirrors check_mwdx_streaming"""
        close = test_data['close']
        factor = 0.2


        close_array = np.array(close, dtype=np.float64)
        batch_result = mwdx(close_array, factor)


        stream = MwdxStream(factor)
        stream_results = []

        for price in close:
            result = stream.update(price)
            stream_results.append(result)


        assert len(stream_results) == len(close)


        for i, val in enumerate(stream_results):
            assert np.isfinite(val), f"NaN at index {i}"


        assert_close(stream_results[0], close[0], atol=1e-9, msg="First value mismatch")


    def test_mwdx_all_nan_input(self):
        """Test MWDX with all NaN values"""
        all_nan = np.full(100, np.nan, dtype=np.float64)


        result = mwdx(all_nan, 0.2)
        assert np.all(np.isnan(result))

    def test_mwdx_batch_performance(self, test_data):
        """Test that batch computation works correctly (performance is secondary)"""
        close = np.array(test_data['close'][:1000], dtype=np.float64)


        batch_result = mwdx_batch(
            close,
            (0.1, 0.9, 0.2)
        )


        assert batch_result['values'].shape == (5, len(close))


        individual_result = mwdx(close, 0.1)
        batch_first = batch_result['values'][0]
        assert_close(batch_first, individual_result, atol=1e-9,
                    msg="MWDX batch mismatch for factor=0.1")

    def test_mwdx_edge_cases(self):
        """Test MWDX with edge case inputs"""

        data = np.arange(1.0, 101.0, dtype=np.float64)
        result = mwdx(data, 0.2)
        assert len(result) == len(data)
        assert np.all(np.isfinite(result))


        data = np.array([50.0] * 100, dtype=np.float64)
        result = mwdx(data, 0.2)
        assert len(result) == len(data)


        for i in range(10, len(result)):
            assert_close(result[i], 50.0, atol=1e-6, msg=f"Constant value failed at index {i}")


        data = np.array([10.0, 20.0, 10.0, 20.0] * 25, dtype=np.float64)
        result = mwdx(data, 0.5)
        assert len(result) == len(data)
        assert np.all(np.isfinite(result))

    def test_mwdx_high_factor(self, test_data):
        """Test MWDX with high factor value"""
        close = np.array(test_data['close'][:100], dtype=np.float64)


        result = mwdx(close, 0.95)
        assert len(result) == len(close)
        assert np.all(np.isfinite(result))

    def test_mwdx_low_factor(self, test_data):
        """Test MWDX with low factor value"""
        close = np.array(test_data['close'][:100], dtype=np.float64)


        result = mwdx(close, 0.01)
        assert len(result) == len(close)
        assert np.all(np.isfinite(result))

    def test_mwdx_consistency(self, test_data):
        """Test that MWDX produces consistent results across multiple calls"""
        close = np.array(test_data['close'][:100], dtype=np.float64)

        result1 = mwdx(close, 0.2)
        result2 = mwdx(close, 0.2)

        assert_close(result1, result2, atol=1e-15, msg="MWDX results not consistent")

    def test_mwdx_step_precision(self):
        """Test batch with various step sizes"""
        data = np.arange(1, 51, dtype=np.float64)


        batch_result = mwdx_batch(
            data,
            (0.2, 0.8, 0.3)
        )

        assert batch_result['values'].shape == (3, len(data))
        expected_factors = [0.2, 0.5, 0.8]
        assert_close(batch_result['factors'], expected_factors, atol=1e-9, msg="Factor mismatch")

    def test_mwdx_batch_error_handling(self):
        """Test MWDX batch error handling"""

        empty = np.array([], dtype=np.float64)
        with pytest.raises(ValueError, match="No input data was provided"):
            mwdx_batch(empty, (0.2, 0.2, 0))


        data = np.random.randn(100).astype(np.float64)
        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            mwdx_batch(data, (0.0, 0.5, 0.1))

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            mwdx_batch(data, (-0.1, 0.5, 0.1))

    def test_mwdx_zero_copy_verification(self, test_data):
        """Verify MWDX uses zero-copy operations"""
        close = np.array(test_data['close'][:100], dtype=np.float64)


        result = mwdx(close, 0.2)
        assert len(result) == len(close)


        batch_result = mwdx_batch(close, (0.1, 0.5, 0.2))
        assert batch_result['values'].shape[0] == 3
        assert batch_result['values'].shape[1] == len(close)

    def test_mwdx_stream_error_handling(self):
        """Test MWDX stream error handling"""

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            MwdxStream(0.0)

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            MwdxStream(-0.5)

        with pytest.raises(ValueError, match="Factor must be greater than 0"):
            MwdxStream(np.nan)

    def test_mwdx_nan_prefix(self):
        """Test MWDX with NaN prefix in data"""

        data = np.array([np.nan, np.nan, 10.0, 20.0, 30.0, 40.0], dtype=np.float64)

        result = mwdx(data, 0.2)
        assert len(result) == len(data)











        assert np.isnan(result[0])
        assert np.isnan(result[1])


        assert result[2] == 10.0
        assert np.abs(result[3] - 12.0) < 1e-10
        assert np.abs(result[4] - 15.6) < 1e-10
        assert np.abs(result[5] - 20.48) < 1e-10


        clean_data = np.array([10.0, 20.0, 30.0, 40.0], dtype=np.float64)
        clean_result = mwdx(clean_data, 0.2)
        assert np.all(np.isfinite(clean_result))

    def test_mwdx_formula_verification(self):
        """Verify MWDX formula: out[i] = fac * data[i] + (1 - fac) * out[i-1]"""
        data = np.array([10.0, 20.0, 30.0, 40.0], dtype=np.float64)
        factor = 0.3

        result = mwdx(data, factor)


        expected = [data[0]]
        for i in range(1, len(data)):
            val = factor * data[i] + (1 - factor) * expected[i-1]
            expected.append(val)

        assert_close(result, expected, atol=1e-12, msg="Formula verification failed")

    def test_mwdx_kernel_parameter(self, test_data):
        """Test MWDX with kernel parameter"""
        close = np.array(test_data['close'][:100], dtype=np.float64)


        for kernel in [None, 'scalar']:
            result = mwdx(close, 0.2, kernel=kernel)
            assert len(result) == len(close)
            assert np.all(np.isfinite(result)), f"NaN values with kernel={kernel}"


        batch_result = mwdx_batch(close, (0.1, 0.3, 0.1), kernel='scalar')
        assert batch_result['values'].shape == (3, len(close))


        with pytest.raises(ValueError, match="Unknown kernel"):
            mwdx(close, 0.2, kernel='invalid')

    def test_mwdx_kernel_consistency(self, test_data):
        """Test that different kernels produce consistent results"""
        close = np.array(test_data['close'][:1000], dtype=np.float64)


        result_auto = mwdx(close, 0.2, kernel=None)
        result_scalar = mwdx(close, 0.2, kernel='scalar')


        assert_close(result_auto, result_scalar, rtol=1e-14,
                    msg="Kernel results differ significantly")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_mwdx.js
/**
 * WASM binding tests for MWDX indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('MWDX partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.mwdx_js(close, 0.2);
    assert(result instanceof Float64Array);
    assert.strictEqual(result.length, close.length);
});

test('MWDX accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.mwdx;
    
    const result = wasm.mwdx_js(
        close,
        expected.defaultParams.factor
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-7,
        "MWDX last 5 values mismatch"
    );
    
    
    await compareWithRust('mwdx', result, 'close', expected.defaultParams);
});

test('MWDX zero factor', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.mwdx_js(inputData, 0.0);
    }, /Factor must be greater than 0/);
});

test('MWDX negative factor', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.mwdx_js(inputData, -0.5);
    }, /Factor must be greater than 0/);
});

test('MWDX NaN factor', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.mwdx_js(inputData, NaN);
    }, /Factor must be greater than 0/);
});

test('MWDX very small dataset', () => {
    
    const data = new Float64Array([42.0]);
    
    const result = wasm.mwdx_js(data, 0.2);
    assert.strictEqual(result.length, 1);
    assert.strictEqual(result[0], 42.0);
});

test('MWDX empty input', () => {
    
    const dataEmpty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.mwdx_js(dataEmpty, 0.2);
    }, /No input data was provided/);
});

test('MWDX reinput', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const firstResult = wasm.mwdx_js(close, 0.2);
    
    
    const secondResult = wasm.mwdx_js(firstResult, 0.3);
    
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    for (let i = 0; i < secondResult.length; i++) {
        assert(isFinite(secondResult[i]), `NaN found at index ${i}`);
    }
});

test('MWDX NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.mwdx_js(close, 0.2);
    
    assert.strictEqual(result.length, close.length);
    
    
    for (let i = 0; i < result.length; i++) {
        assert(isFinite(result[i]), `NaN found at index ${i}`);
    }
});

test('MWDX batch', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const batch_result = wasm.mwdx_batch_js(
        close, 
        0.1, 0.5, 0.1    
    );
    
    
    const rows_cols = wasm.mwdx_batch_rows_cols_js(0.1, 0.5, 0.1, close.length);
    const rows = rows_cols[0];
    const cols = rows_cols[1];
    
    assert(batch_result instanceof Float64Array);
    assert.strictEqual(rows, 5); 
    assert.strictEqual(cols, close.length);
    assert.strictEqual(batch_result.length, rows * cols);
    
    
    const individual_result = wasm.mwdx_js(close, 0.1);
    const batch_first = batch_result.slice(0, close.length);
    
    assertArrayClose(batch_first, individual_result, 1e-9, 'MWDX first combination');
});

test('MWDX different factors', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const testFactors = [0.1, 0.2, 0.5, 0.9];
    
    for (const factor of testFactors) {
        const result = wasm.mwdx_js(close, factor);
        assert.strictEqual(result.length, close.length);
        
        
        let finiteCount = 0;
        for (let i = 0; i < result.length; i++) {
            if (isFinite(result[i])) finiteCount++;
        }
        assert.strictEqual(finiteCount, close.length, 
            `Found NaN values for factor=${factor}`);
    }
});

test('MWDX batch performance', () => {
    
    const close = new Float64Array(testData.close.slice(0, 1000)); 
    
    
    const startBatch = performance.now();
    const batchResult = wasm.mwdx_batch_js(
        close,
        0.1, 0.9, 0.2    
    );
    const batchTime = performance.now() - startBatch;
    
    
    const metadata = wasm.mwdx_batch_metadata_js(0.1, 0.9, 0.2);
    
    const startSingle = performance.now();
    const singleResults = [];
    
    for (const factor of metadata) {
        const result = wasm.mwdx_js(close, factor);
        singleResults.push(...result);
    }
    const singleTime = performance.now() - startSingle;
    
    
    console.log(`Batch time: ${batchTime.toFixed(2)}ms, Single time: ${singleTime.toFixed(2)}ms`);
    
    
    assertArrayClose(batchResult, singleResults, 1e-9, 'Batch vs single results');
});

test('MWDX edge cases', () => {
    
    
    
    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = i + 1;
    }
    const result = wasm.mwdx_js(data, 0.2);
    assert.strictEqual(result.length, data.length);
    
    
    for (let i = 0; i < result.length; i++) {
        assert(isFinite(result[i]), `NaN at index ${i}`);
    }
    
    
    const constantData = new Float64Array(100).fill(50.0);
    const constantResult = wasm.mwdx_js(constantData, 0.2);
    
    assert.strictEqual(constantResult.length, constantData.length);
    
    
    for (let i = 10; i < constantResult.length; i++) {
        assertClose(constantResult[i], 50.0, 1e-6, 
            `Constant value failed at index ${i}`);
    }
});

test('MWDX batch metadata', () => {
    
    const metadata = wasm.mwdx_batch_metadata_js(
        0.2, 0.6, 0.2    
    );
    
    
    assert.strictEqual(metadata.length, 3);
    assertClose(metadata[0], 0.2, 1e-9);
    assertClose(metadata[1], 0.4, 1e-9);
    assertClose(metadata[2], 0.6, 1e-9);
});

test('MWDX consistency across calls', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    const result1 = wasm.mwdx_js(close, 0.2);
    const result2 = wasm.mwdx_js(close, 0.2);
    
    assertArrayClose(result1, result2, 1e-15, "MWDX results not consistent");
});

test('MWDX step precision', () => {
    
    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1;
    }
    
    const batch_result = wasm.mwdx_batch_js(
        data,
        0.2, 0.8, 0.3     
    );
    
    
    const rows_cols = wasm.mwdx_batch_rows_cols_js(0.2, 0.8, 0.3, data.length);
    const rows = rows_cols[0];
    
    
    assert.strictEqual(rows, 3);
    assert.strictEqual(batch_result.length, 3 * data.length);
    
    
    const metadata = wasm.mwdx_batch_metadata_js(0.2, 0.8, 0.3);
    assert.strictEqual(metadata.length, 3);
    assertClose(metadata[0], 0.2, 1e-9);
    assertClose(metadata[1], 0.5, 1e-9);
    assertClose(metadata[2], 0.8, 1e-9);
});

test('MWDX streaming simulation', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    const factor = 0.2;
    
    
    const batchResult = wasm.mwdx_js(close, factor);
    
    
    assert.strictEqual(batchResult.length, close.length);
    
    
    for (let i = 0; i < batchResult.length; i++) {
        assert(isFinite(batchResult[i]), `Expected finite value at index ${i}`);
    }
    
    
    assertClose(batchResult[0], close[0], 1e-9, "First value mismatch");
});

test('MWDX high factor', () => {
    
    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = Math.sin(i * 0.1) * 100 + 1000; 
    }
    
    const result = wasm.mwdx_js(data, 0.95);
    
    assert.strictEqual(result.length, data.length);
    
    
    for (let i = 0; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }
});

test('MWDX low factor', () => {
    
    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = Math.sin(i * 0.1) * 100 + 1000; 
    }
    
    const result = wasm.mwdx_js(data, 0.01);
    
    assert.strictEqual(result.length, data.length);
    
    
    for (let i = 0; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }
});

test('MWDX NaN prefix', () => {
    
    const data = new Float64Array([NaN, NaN, 10.0, 20.0, 30.0, 40.0]);
    
    const result = wasm.mwdx_js(data, 0.2);
    assert.strictEqual(result.length, data.length);
    
    
    
    
    
    
    
    
    
    
    
    assert(isNaN(result[0]), 'Expected NaN at index 0');
    assert(isNaN(result[1]), 'Expected NaN at index 1');
    
    
    assert.strictEqual(result[2], 10.0);
    assert(Math.abs(result[3] - 12.0) < 1e-10, `Expected 12.0 at index 3, got ${result[3]}`);
    assert(Math.abs(result[4] - 15.6) < 1e-10, `Expected 15.6 at index 4, got ${result[4]}`);
    assert(Math.abs(result[5] - 20.48) < 1e-10, `Expected 20.48 at index 5, got ${result[5]}`);
    
    
    const cleanData = new Float64Array([10.0, 20.0, 30.0, 40.0]);
    const cleanResult = wasm.mwdx_js(cleanData, 0.2);
    
    
    for (let i = 0; i < cleanResult.length; i++) {
        assert(isFinite(cleanResult[i]), `Expected finite value at index ${i} for clean data`);
    }
});

test('MWDX formula verification', () => {
    
    const data = new Float64Array([10.0, 20.0, 30.0, 40.0]);
    const factor = 0.3;
    
    const result = wasm.mwdx_js(data, factor);
    
    
    const expected = [data[0]]; 
    for (let i = 1; i < data.length; i++) {
        const val = factor * data[i] + (1 - factor) * expected[i-1];
        expected.push(val);
    }
    
    assertArrayClose(result, expected, 1e-12, 'Formula verification failed');
});

test('MWDX all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        allNaN[i] = NaN;
    }
    
    
    const result = wasm.mwdx_js(allNaN, 0.2);
    for (let i = 0; i < result.length; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i}`);
    }
});

test('MWDX oscillating data', () => {
    
    const data = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        data[i] = (i % 2 === 0) ? 10.0 : 20.0;
    }
    
    const result = wasm.mwdx_js(data, 0.5);
    assert.strictEqual(result.length, data.length);
    
    
    for (let i = 0; i < result.length; i++) {
        assert(isFinite(result[i]), `Expected finite value at index ${i}`);
    }
});

test('MWDX small step size', () => {
    
    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1;
    }
    
    const batch_result = wasm.mwdx_batch_js(
        data,
        0.1, 0.2, 0.05     
    );
    
    const rows_cols = wasm.mwdx_batch_rows_cols_js(0.1, 0.2, 0.05, data.length);
    const rows = rows_cols[0];
    
    assert.strictEqual(rows, 3);
    assert.strictEqual(batch_result.length, 3 * data.length);
});